Code review changes

ZachAttackE opened the pull request:
Pull request for updates to airbrake program, based off of code review for the recovery program, and adding control algorithm for the airbrakes.

ZachAttackE committed:
updated variables

updated variables to be more descriptive and other changes.

ZachAttackE committed:
Made changes similar to recovery changes

made the file not write to if it is not open. Made the computer not arm if it is moving. Made detect_takeoff same as recovery computer.

ZachAttackE committed:
added target acceleration algorithm

added the algorithm to find the target acceleration in order to reach the target altitude.

ZachAttackE committed:
added p controller

added p controller algorithm to control the airbrakes to reach the target acceleration.

ZachAttackE committed:
adjusted log file header

adjusted log file header and updated some comments. Removed full brake command at MCO. The predict_apogee function will set the brakes.

ZachAttackE committed:
Fixed bug in barometer reading

bug prevented the reading to be validated.

ZachAttackE committed:
Adjusted Read me

adjusted the read me information

ZachAttackE committed:
logging update

updated the header format

ZachAttackE committed:
updated brake position var

updated brake position variable to 0 once they are closed at apogee.

ZachAttackE committed:
comment update

ZachAttackE committed:
maximum braking velocity

Made it to where the brakes wouldn't be actuated if the velocity is too high.

ZachAttackE committed:
updated read me

ZachAttackE committed:
Adjusted P-term

Added a P-term algorithm based off of velocity to keep the brakes responsive as the rocket slows down.

linux-colonel commented on WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES/WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES.ino:
```diff
@@ -67,30 +67,63 @@ READ ME
 
```
Need versions for libraries.

linux-colonel commented on WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES/WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES.ino:
```diff
@@ -13,25 +13,25 @@ READ ME
   Set Current Location and Time Sea Level Pressure:
```
This section should go into a  `README.md`  file, similar to what's decribed [here](https://github.com/Palouse-Aerospace-2023-Spaceport-Cup/SRAD-Deployment-Computer/pull/2#discussion_r1096875592)

linux-colonel commented on WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES/WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES.ino:
```diff
@@ -67,30 +67,63 @@ READ ME
+  #define TAKEOFF_ALTITUDE  (1) //altitude above the ground in meters that triggers the detect_take_off function
+
+// *********SET ACCELERATION OFFSET CONSTANT HERE***************
+  #define OFFSET (-7)// offset for acceleration data in m/(s^2) for prediction algorithm
```
How was this value derived?

linux-colonel commented on WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES/WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES.ino:
```diff
@@ -67,30 +67,63 @@ READ ME
   
 //**************SET TARGET ALTITUDE HERE*************************
-  float target_altitude = 100; //target altitude above ground in meters
+  #define TARGET_ALTITUDE (1) //target altitude above ground in meters
```
did you mean to change this value from 100 meters to 1 meter?

linux-colonel commented on WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES/WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES.ino:
```diff
@@ -67,30 +67,63 @@ READ ME
+  #define TARGET_ALTITUDE (1) //target altitude above ground in meters
 
 // *********SET SEA LEVEL PRESSURE HERE***************
   #define SEALEVELPRESSURE_HPA (1016.00)//set according to location and date
```
some of these should go in configuration files:
* `TARGET_ALTITUDE`
* `SEALEVELPRESSURE_HPA`
* `hz`
* `TAKEOFF_ALTITUDE`
* `OFFSET`

you can re-use the code you wrote for the deployment computer by putting it in a library.  This will require a separate git repo.  I can help you get this set up.

linux-colonel commented on WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES/WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES.ino:
```diff
@@ -67,30 +67,63 @@ READ ME
   
 // *********SET FREQUENCY HERE IN HZ***************
-  #define hz (5)//hz
+  #define hz (10)//hz
```
Why the `()` around all the values that are `#define`?

linux-colonel commented on WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES/WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES.ino:
```diff
@@ -120,28 +153,19 @@ void setup() {
+  close_brakes();
+  
   /*
   //open & close brakes three times to identify issues
```
If you do want to open/close 3 times, a for loop would be appropriate here.

linux-colonel replied:
would it be a safety issue if the rocket did this while it was on the pad?  I imagine the brakes opening and knocking the person arming the electronics off the ladder...

linux-colonel commented on WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES/WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES.ino:
```diff
@@ -195,128 +228,138 @@ void setup() {
-  myFile.print(F("Liftoff Detected"));
 
 
+  if (logFile) {
```
This block can go in a function, similar to [this](https://github.com/Palouse-Aerospace-2023-Spaceport-Cup/SRAD-Deployment-Computer/pull/2#discussion_r1096883486)

linux-colonel commented on WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES/WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES.ino:
```diff
@@ -195,128 +228,138 @@ void setup() {
+//DESCENDING MODE********************
 
+  close_brakes();//fully closes brakes
+  brake_position = 0;
```
Should this go in the `close_brakes()` function?

linux-colonel commented on WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES/WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES.ino:
```diff
@@ -373,70 +416,71 @@ void read_barometer(){//takes barometer reading
 }
 
 float read_altitude(){//reads and returns barometer altitude reading
```
This would be a candidate for a library.  We can discuss.

linux-colonel replied:
same with the `read_barometer()` function

linux-colonel commented on WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES/WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES.ino:
```diff
@@ -373,70 +416,71 @@ void read_barometer(){//takes barometer reading
 
-void set_brakes(float pos){//set brakes to any position between 0 and 1 (0 = closed, 1 = open) 
-  brakes.write((int)(pos*(brakes_open - brakes_closed) + brakes_closed));//600 to 1580 for full range
+void set_brakes(){//set brakes to desired brake position between 0 and 100 (0 = closed, 100 = open) 
```
This function should take a parameter `pos` for the position, rather than using a global variable.  It makes it much clearer what's being set.  Global variables should be used sparingly.

linux-colonel commented on WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES/WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES.ino:
```diff
@@ -67,30 +67,63 @@ READ ME
   
-  //brakes servo RANGE
+  //brakes servo RANGE and Velocity Limit
   #define brakes_closed  625
```
These three lines should be in all CAPS since they are constants.  It's convention.

linux-colonel commented on WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES/WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES.ino:
```diff
@@ -67,30 +67,63 @@ READ ME
-  #define G (6.3)// m/(s^2)
 
 //Brakes Servo Handle
   Servo brakes; 
```
This comment is actually for line 99/131 below.  What is the comment next to the `mpu` object mean?

linux-colonel commented on WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES/WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES.ino:
```diff
@@ -373,70 +416,71 @@ void read_barometer(){//takes barometer reading
+  brakes.write(brakes_open);
 }
 
 void close_brakes(){//opens brakes fully
```
comment is redundant - the name of the function makes it obvious as what it does.

linux-colonel commented on WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES/WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES.ino:
```diff
@@ -453,61 +497,112 @@ void read_accelerometer(){//reads and updates z_global variable
   acc = 9.81 *  -(1+mpu.getAccX()*sin(mpu.getAngleY()*3.14/180)-mpu.getAccY()*cos(mpu.getAngleY()*3.14/180)*sin(mpu.getAngleX()*3.14/180)-mpu.getAccZ()*cos(mpu.getAngleY()*3.14/180)*cos(mpu.getAngleX()*3.14/180));// uses a transformation matrix to determine global z acceleration
```
I'd put the full mathematical formula in a comment, one line above.

linux-colonel commented on WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES/WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES.ino:
```diff
@@ -453,61 +497,112 @@ void read_accelerometer(){//reads and updates z_global variable
+  distance_to_target = TARGET_ALTITUDE - x_current; //update distance to target. Negative value implies target has been passed.
+}
+
+void predict_apogee(){//predicts apogee with current velocity, altitude, and acceleration and sets brakes to reach target altitude
```
I recommend renaming this function to something that makes it obvious that the air brakes are deployed here.  Maybe `predict_and_deploy()`.  I'm not the best with naming.

linux-colonel commented on WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES/WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES.ino:
```diff
@@ -453,61 +497,112 @@ void read_accelerometer(){//reads and updates z_global variable
-  apo = - sq(vel)/(2*(acc - G)) * log((sq(vel) + sq(vel) * (-G)/(acc - G))/(sq(vel)*(-G)/(acc -G))) + x2;
+
+
+int detect_take_off(){//returns 1 if take off is detected, otherwise 0
```
`bool` return type

linux-colonel commented on WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES/WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES.ino:
```diff
@@ -453,61 +497,112 @@ void read_accelerometer(){//reads and updates z_global variable
 
-int MCO(){//returns 1 if MCO is detected, 0 otherwise.
+
+int detect_mco(){//returns 1 if MCO is detected, 0 otherwise.
```
`bool` return type

linux-colonel commented on WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES/WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES.ino:
```diff
@@ -453,61 +497,112 @@ void read_accelerometer(){//reads and updates z_global variable
 }
 
+
 int detect_apogee(){//looks for apogee and returns 1 if detected, 0 if not.
```
`bool` return type

linux-colonel commented on WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES/WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES.ino:
```diff
@@ -453,61 +497,112 @@ void read_accelerometer(){//reads and updates z_global variable
+  if(counter_apogee > 3){    //enters if last 3 positions are lower than the previous positions
     return 1;   //returns 1 for apogee detected
   }
   else{           
```
style: `} else {` should be on one line

linux-colonel commented on WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES/WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES.ino:
```diff
@@ -518,13 +613,13 @@ int detect_apogee(){//looks for apogee and returns 1 if detected, 0 if not.
 
 int detect_touchdown(){//returns 1 if touchdown is detected, otherwise returns 0.
```
`bool` return type

linux-colonel commented on WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES/WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES.ino:
```diff
@@ -518,13 +613,13 @@ int detect_apogee(){//looks for apogee and returns 1 if detected, 0 if not.
+    counter_landed = 0;    //resets counter
     return 1; //returns 1 if touchdown detected
   }
   else{
```
style: `} else {` should be on one line.

linux-colonel reviewed (changes requested):
Again, overall looks good!  Got a bunch of nit-picky stuff, but it shouldn't take long to fix.  Please let me know if you have any questions.  I can meet with you Thursday or Friday night if you want.

ZachAttackE committed:
Adjust set positions

Adjusted open and close positions for crimson arrow. Also set target altitude to 100,000m so brakes will not deploy. This is for test flight at Brother Oregon.

ZachAttackE committed:
fixed apogee bug

fixed bug that predicted apogee to early. Also adjusted p-controller values for crimson arrow and set target altitude to 10k ft.

ZachAttackE committed:
Update WINGSTOP_Air_Brake_Flight_Computer_FULL_BRAKES.ino
